specification.md

```diff
@@ -439,7 +439,7 @@ The second way would be to directly filter only the "veggie like" vegetables wit
 | fromStateData | Workflow expression that filters state data that can be used by the action | string | no |
 | useResults | If set to `false`, action data results are not added/merged to state data. In this case 'results' and 'toStateData' should be ignored. Default is `true`.  | boolean | no |
 | results | Workflow expression that filters the actions data results | string | no |
-| toStateData | Workflow expression that selects a state data element to which the action results should be added/merged into. If not specified denotes the top-level state data element | string | no |
+| toStateData | Workflow expression that selects a state data element to which the action results should be added/merged into. If not specified denotes the top-level state data element. In case it is not specified and the result of the action is not an object, that result should be merged as value of `response` key. If that `response` key already exist in the model, its value will be overwritten. | string | no |
```

Comment:
@fjtirado I had totally forgotten about #712... 
Don't you think it's better to settle on something like: if you dont set `toStateDate`, the output of an action is simply ignored? 
I'm not really confortable with this arbitrary decision to enforce a default behavior for this when not specified.
I personally would just remove: `If not specified denotes the top-level state data element` and instead indicate that if not set, nothing happens, which is IMHO more intuitive.
@JBBianchi @ricardozanini @fjtirado What do you guys think?

Reply:
hmmmm, I think the non object should still be merged (the model has to be always an object so setting it as the flow model is not an option). The discussion, I think, is to decide the default name of the property to which the non object value is assigned- 
I propose `response` because thats the one that is being currently used in our implementation. Might be not best name, as for example `key` for the expression used to generate the workflow instance id, but it is not the worse either. As an enhacement, we can add a property at the header of the workflow definition to specify the name (if response is not the one that is wanted) to enforce implementors to support that.

Reply:
That's not my point. Object or non-object is irrelevant to the matter I'm speaking about. My concern is the fact that, if the `toStateData` is not set, the output should not be assigned to anything, it should instead be ignored, which I believe is the (intuitive) way to go.

Reply:
Maybe I am influenced by the current behaviour (merging always) but I found it pretty intuitive. Most actions produce a result and that result should be available to the model. If the result of the action is irrelevant (for an action that actually produce a result, because tipically actios which result is not relevan will not produce any result, so nothing is merged), In many case it is ok to just the result (object or not object) direcly into the model, without the need to perform any key traslation. 
If we force people to write toStateData to perform a merge, the resulting definition will become too verbose, in my opinion. Im actually surprised you are proposing this ;)

Reply:
Thats said, since we already has customers, maybe we should request feedback about this to them. But I think this will imply them to write toStateData for every rest request they perform, when they are certainly fine with the current merge approach. 
Maybe we can just leave merge as default and globally change that by an optional key in the header, so we have an optional `mergePolicy` which default value is merge, but that can be none.

Reply:
I understand your point of view and it does make sense IMO.
However, my problem is with the "tricks" that have been put into place to avoid having to automatically merge, meaning the (very ugly) `useResult` property. The latter is not mutually exclusive to `toStateData`, and is therefore leaving the door opened to wierd definitions where both properties are set, even though they address the same concern.
My idea is to remove the `useResult` property, and give the feature to an empty `toStateData`, but I agree that, in that case, the workflow might become a bit more verbose.
I believe that the issue is related to the concept of `stateData`: what is it and how should it be used? If it's a concept that represents contextual data at a point in time (aka state) of a workflow's execution, then I believe it should never be overidden by an action's output. Furthermore, there's a conceptual issue where the output of an action should be implicitly used as the input of next action, save if the `fromStateData` has been used. This always felt wierd to me, and to our clients, too, even though it can be practical in exotic use cases. In my opinion, `stateData` should not be changed by actions if not explicitly requested, and actions should always receive the output, if any, of the previous action, thus making `fromStateData` useless (which it is, providing you could access state data with something similar to a `$STATE` expression parameter).

Reply:
> Thats said, since we already has customers, maybe we should request feedback about this to them

Yes, you are totally right, I believe this very subtle issue is actually bigger than us, and should be submitted to our community and opposed to their factual use cases.

Reply:
As as summary of my previous comment, as alternative proposal
1) Remove `actionDataFilter`. Add actionResultFilter which basically modified at will the result of the action execution
2) This result is not merged into the workflow model, but make accessible to the next action (or to the state if it is the last acion) as `$WORKFLOW.lastActionResult`. This is also accesible to `stateDataFilter.output`. Note that lastActionResult might be json object or a non object, it does not really matter
3) Add a new mandatory/optional (point of discussion) property to operation state. `resultPolicy`, posible values are `merge` (lastActionResult is merged into the model)  or `ignore` (do nothing)
4) When merging a non object into the workflow model (for any state), that non object will be associated with `response` key. That `response` key might be change by specifyng another value through workflow level property `nonObjectm̀ergeKey`

Reply:
I think that `resultPolicy`should be mutually exclusive with `stateDataFilter.ouptut`. I mean, either the user specified  how to merge `lastActionResult` into the model by setting and output filter of the state or he specifies that `lastActionResult` is to be merged as it is or that nothing should be done
Another way to look at it is that resultPolicy has three posible values: filter, merge o none. 
If you specify the filter you can assume its value is filter. Then if you do not say anything, it can be assumed (for continuity with previous version) that you want to merge the result. And only if you do not want to do anything, you specify `ignore` (thats the boolean we currently have, that in any case will be moved to the right place, the state) I think is fine to just an optional boolean "doNotMerge" to operation state. I think 99% of the time you want your operation state to merge the result (that was not true for the action, thats why moving the boolean to state level does the trick)

Reply:
@cdavernas wdyt?

Reply:
> Remove actionDataFilter. Add,as replacement actionResultFilter of type string

Good suggestion. However, this removes the user's ability to add potentially critical data to the state's after executing an action: only the last action's output is made available. There are several use cases where that is problematic and too restrictive.

Reply:
For a short term fix, I think we should indicate that when the `toStateData` property of an `actionDataFilter` is not set, it does not output anything to the state's data.

For the `stateDataFilter`, however, we should proceed as initially planned, meaning that if not set, it should be merged back at top level. If the data is not an object, it should be merged back into a field such as `response`, as you suggested, though I'm not a fan of that specific term. Why not simply calling it `output`, which is more ubiquitous IMHO than response, which implies the concept of a request?

Reply:
For a longer term solution, I think we should first rename `dataFilter`. The term is not ubiquitous, and might even be confusing to some users.

I would personally rename it to `output`, and modifiy it so that it looks like:

```yaml
actions:

- name: action-with-no-output
  output: false

- name: action-with-filtered-output-1
  output: '${ .myFilteredProperty }'

- name: action-with-filtered-output-2
  output:
    filter: '${ .myFilteredProperty }'

- name: action-that-sets-state-data
  output:
    to: '${  .myStateDataProperty }'
```

The same could be done for `input`. And we could even include the (optional) schema in those new objects.

We could also allow action/state data to be of any kind (object, array, primitive), as I don't see the need to enforce it to be an object, which in addition force us to propose ways around it in specific use cases.

Reply:
@cdavernas It seems we do not agree on default behaviour over the response of the action. I think that, by default, the result of the last action of an operation state should be merged into the workflow model. This is the current default and is a good one. As a short term fix, all we need to do is to indicate what to do when the result is not an object (my proposal is to use a configurable key, which defaults to "response")

Reply:
One possibility is to merge automatically the result of the last action and only merge the result of intermediate actions if there is a toStateData or similar. Most typical use case is an operation with just one action that invokes a rest (or other function returning a value, json object or not), for that use case, that will work smoothly. For action chains, I agree that only allowing merging the result of the last action is too restrictive. Therefore, for intermediate actions, your sugesstion of only merging when the user explicitly states it makes perfect sense.

Reply:
But we are discussing two different things:
1) To merge o nor to merge
2) How to merge a non object json. 
I propose, within the scope of this PR, just discuss number 2)

Reply:
Agreed!

For 2, I instead propose using `output`, that is IMHO more ubiquitous.

Reply:
Ok, since action name is now mandatory, we can enforce <actionName>Output, wdyt?

Reply:
Done

Reply:
Opened https://github.com/serverlessworkflow/specification/issues/837 to follow up this